cram/cram_decode.c

```diff
@@ -2896,24 +2895,22 @@ static int cram_to_bam(sam_hdr_t *sh, cram_fd *fd, cram_slice *s,
         qual = NULL;
     }
 
-    bam_idx = bam_construct_seq(bam, cr->aux_size + rg_len,
-                                name, name_len,
-                                cr->flags,
-                                cr->ref_id,
-                                cr->apos,
-                                cr->aend,
-                                cr->mqual,
-                                cr->ncigar, &s->cigar[cr->cigar],
-                                cr->mate_ref_id,
-                                cr->mate_pos,
-                                cr->tlen,
-                                cr->len,
-                                seq,
-                                qual);
-    if (bam_idx == -1)
-        return -1;
+    bam_idx = bam_set1(*bam,
+                       name_len, name,
+                       cr->flags, cr->ref_id, cr->apos - 1, cr->mqual,
+                       cr->ncigar, &s->cigar[cr->cigar],
+                       cr->mate_ref_id, cr->mate_pos - 1, cr->tlen,
+                       cr->len, seq, qual,
+                       cr->aux_size + rg_len);
+    if (bam_idx < 0) {
+        return bam_idx;
+    }
+
+    // mark the buffer space allocated for aux data as containing actual aux data
+    (*bam)->l_data += cr->aux_size + rg_len;
```

Comment:
Better to increment `…->l_data` individually in the two `if` blocks below so they're self-contained and if a third item is added later the changes needed are more self-contained.

On successful return from `bam_set1()`, `bam_idx` is the same as `…->l_data` and the incrementing maintains that. So IMHO you should change the `return` below to `return …->l_data` so there's no need to update `bam_idx`. (And then you can rename `bam_idx` to `ret` and just use it to check the return value from `bam_set1()` — I have no idea what the name `bam_idx` in the original code was trying to imply!).

Reply:
ok!